grr/server/grr_response_server/gui/ui/lib/queued_exhaust_map.ts

```diff
@@ -0,0 +1,69 @@
+import { ObservableInput, OperatorFunction, ObservedValueOf, pipe, EMPTY } from 'rxjs';
+import { concatMap, tap, map } from 'rxjs/operators';
+
+class NotImplementedError extends Error { };
+
+interface RequestTracker<T> {
+  recordLastTag(tag: T): void;
+  shouldProcess(thisTag: T): boolean;
+}
+
+class SingleIndexTracker implements RequestTracker<number> {
+  private lastIndex = -1;
+
+  recordLastTag(tag: number): void {
+    this.lastIndex = tag;
+  }
+
+  shouldProcess(thisTag: number): boolean {
+    return this.lastIndex === thisTag;
+  }
+}
+
+class ValueWithTag<V, T> {
+  constructor(
+    readonly value: V,
+    readonly tag: T,
+  ) { }
+}
+
+function tagByIndex<V>(indexTracker: SingleIndexTracker) {
+  return pipe(
+    map((value: V, index) => new ValueWithTag(value, index)),
+    tap((valueWithTag: ValueWithTag<V, number>) => {
+      indexTracker.recordLastTag(valueWithTag.tag);
+    }),
+  );
+}
+
+/**
+ * Similar to exhaustMap, but buffers the specified number of most recent source values
+ * in a queue of limited size. The source values that are not discarded are projected to
+ * Observables and merged in the output Observable in a serialized fashion, waiting for
+ * each one to complete before merging the next.
+ *
+ * Important notes:
+ * - Currently only queue of size 1 is supported.
+ * - If the source produces values faster than this operation discards them, no values will
```

Comment:
Yeap you are right, that's a different issue I forgot to describe, and there is also a warning in the [RxJS concatMap page](https://rxjs-dev.firebaseapp.com/api/operators/concatMap):
> Warning: if source values arrive endlessly and faster than their corresponding inner Observables can complete, it will result in memory issues as inner Observables amass in an unbounded buffer waiting for their turn to be subscribed to.

What about the "If the source produces values faster...": do you think that's a valid note?